src/main/java/org/radarbase/ksql/udaf/InterQuartileRangeUdaf.java

```diff
@@ -0,0 +1,51 @@
+package org.radarbase.ksql.udaf;
+
+import io.confluent.ksql.function.udaf.Udaf;
+import io.confluent.ksql.function.udaf.UdafDescription;
+import io.confluent.ksql.function.udaf.UdafFactory;
+import java.util.ArrayList;
+import java.util.List;
+import org.apache.commons.math3.stat.descriptive.DescriptiveStatistics;
+
+@UdafDescription(name = "iqr",
+        author = "yatharthranjan",
+        version = "1.0.0",
+        description = "Calculates the Inter-Quartile Range of the values in a stream.")
+public class InterQuartileRangeUdaf {
+
+    private InterQuartileRangeUdaf() {
+
+    }
+
+    @UdafFactory(description = "Calculates the Inter-Quartile Range of values in a stream.")
+    public static Udaf<Float, List<Float>, Float> createUdaf() {
+        return new InterQuartileRangeUdafImpl();
+    }
+
+    private static class InterQuartileRangeUdafImpl implements Udaf<Float, List<Float>, Float> {
+        @Override
+        public List<Float> initialize() {
+            return new ArrayList<>();
```

Comment:
For high-frequency streams or large time intervals, this array list will become very large and cause a lot of I/O delays and memory consumption. Use a sampling reservoir to put a limit on the number of samples:
https://github.com/RADAR-base/radar-commons/blob/master/radar-commons-server/src/main/java/org/radarbase/stream/collector/UniformSamplingReservoir.java. You would need to use a UDF `Struct` as the value type since a sampling reservoir needs to keep a total count (number of values represented by the samples) and the samples themselves. For low frequency streams or very small intervals, this approach is fine.

Reply:
Hi, yes i was wondering if we needed the same. Currently, these are mostly for low-frequency Garmin data (from the health API), plus on our kcl private cloud we have loads of ram but very limited compute, but I agree would be good to generally make this less I/O intensive so it works well in all cases. I will look into doing that.

Reply:
Implementation notes: the UniformSamplingReservoir above keeps a sorted list to speed up the measure calculations. If you don't need that because the Apache math tools already does that internally, the implementation can be simplified. When an existing value should be replaced, as determined by the random draw, pick a random index, and replace it with the new value.
```java
public void add(float value) {
        if (samples.size() == maxSize) {
            long replaceIndex = ThreadLocalRandom.current().nextLong(count);
            if (replaceIndex < maxSize) {
                samples.set(replaceIndex, value);
            }
        } else {
            samples.add(value);
        }

        count++;
}
```

Reply:
added now